lib/parse-args.js

```diff
@@ -188,6 +191,28 @@ function buildYargs (withCommands = false) {
   return yargs
 }
 
+function loadConfigFile (path, ext) {
+  let config = {}
+  const jsExts = ['.js', '.cjs']
+  const ymlExts = ['.yml', '.yaml']
+  if (jsExts.includes(ext)) {
+    config = require(path)
+  } else if (ymlExts.includes(ext)) {
+    config = require('js-yaml').load(readFileSync(path, 'utf8'))
+  } else {
+    config = JSON.parse(readFileSync(path, 'utf8'))
+  }
+  return config
+}
+
+function getConfigFileNames () {
+  const files = ['nyc.config.js']
+  const prefixes = ['.c8rc', '.nycrc']
+  const exts = ['.yaml', '.yml', '.js', '.cjs', '.mjs', '.json', '']
+  prefixes.forEach((prefix) => exts.forEach((ext) => files.push(`${prefix}${ext}`)))
```

Comment:
If I'm reading the [find-up docs](https://github.com/sindresorhus/find-up#findupsyncmatcher-options) correctly, the order of this array is the order of precedence if there are multiple config files on disk - the first in the array being the highest priority.
This code therefor creates a priority as follows:
1. `nyc.config.js`
2. `.c8rc.yaml`
3. `.c8rc.yml`
2. `.c8rc.js`
3. `.c8rc.cjs`
3. `.c8rc.mjs`
3. `.c8rc.json`
3. `.c8rc`
&c. into the `.nycrc` varients.

To me this priority list feels odd and looks to be in disagreement with the [precedence set by `nyc`](https://github.com/istanbuljs/load-nyc-config).

Reply:
This is a good suggestion.  

> While I'd love to see mjs support, the catch there is that ESM doesn't like getting loaded into a CommonJS process. The other way 'round can be done easily, but I think it's out of scope for this ticket to convert this package to ESM... :P

From my understanding, and from the comments above, we would like to remove the .mjs requirement as it was out of scope for this particular patch.  I am going to remove the '.mjs' from the ext list.

Reply:
Based on the comment above, I think the acceptable list of file names should be the following in order of priority.

```
 .c8rc
.c8rc.json
.c8rc.yml
.c8rc.yaml
.c8.config.js
.c8.config.cjs
.nycrc
.nycrc.json
.nycrc.yml
.nycrc.yaml
nyc.config.js
nyc.config.cjs
```

Reply:
> Tests that prove that each config file type is being loaded correctly, for every possible config file type. I strongly suspect there will be an error or warning when loading an mjs file.

I have a strategy I would like to suggest to accomplish this goal and get some feedback.  In order to test each configuration file being loading, I would need to run c8 in a directory or subdirectory containing that configuration file.  IMO, I have two options

1.  I could add a folder to the test directory labeled config-test or something similar and manually add subdirectories for the various configuration files. 
2. I could add a before and after hook for each configuration file unit test, and create these subdirectories at run time.  Obviously, the after hook would clean up after the before hook.

 I think there are advantages and drawbacks to each of the options.  

For option 1, this is many files to add to the test directory and there could be possible duplications of files from the config directory.

For option 2, since the directory creation and tear down is at run time, this could potentially lead to false positive test results, files accidentally getting committed to the code base, if the after function doesn't execute correctly, and it can make it more difficult to read what the test case is attempting to accomplish.  

@bcoe and @kf6kjg Do you see any other approaches?  Is there a preference?

Reply:
Or you could use a dependency injection pattern where you can 
1. inject the found file path instead of reading the directory tree, and
2. inject the config you want to test in place of the file read.

This provides better and faster tests since there's no actual disk seeks and you are in 100% full control in each test.

One tool I've found that I like is [tsyringe](https://github.com/microsoft/tsyringe), but often for simple cases I'll just do something like 

```typescript
function loadConfigFile(path, ext, _di = { readJsFile: require, readFile: readFileSync }) {
  // same logic, just replaces the relevant blocks with things like
  config = JSON.parse(_di.readFile(path, 'utf8'));
}

// Or even more complex:

function loadConfigFile(path, ext, _di = { readJsFile: require, readYamlFile: (path) => require('js-yaml').load(readFileSync(path, 'utf8')), readJsonFile: (path) => JSON.parse(readFileSync(path, 'utf8')) }) {
  // same logic, just replaces the relevant blocks with things like
  config = _di.readJsonFile(path);
}
```

With that you can in your tests pass to the function all the overrides you need to prove that the function is behaving correctly - including proving that it is indeed calling the functions.

Reply:
As to the order and file names I prefer my tooling config files, even the JS-based ones, to be hidden files. Thus I promote the following ordered list:

```
.c8rc
.c8rc.json
.c8rc.yml
.c8rc.yaml
.c8rc.js
.c8rc.cjs
.c8.config.js
.c8.config.cjs
c8.config.js
c8.config.cjs
.nycrc
.nycrc.json
.nycrc.yml
.nycrc.yaml
nyc.config.js
nyc.config.cjs
```

Having an `rc` file be a JS file may seem odd to some, but it's a pattern that's supported in ESLint, Mocha, Commitlint, Lint-Staged, Prettier, and more.

Reply:
I just created a list of supported files in a [function](https://github.com/bcoe/c8/pull/436/commits/6823fa2a1f7b25a9386f0b858357e31fc9e00dcc#diff-40d93e4e5f4b94fc8a34a87a3328651c24e855051bfcdd6944dee76f9c82b8d4R217-R238)

In terms of [test cases](https://github.com/bcoe/c8/pull/436#discussion_r1113740681), I created test cases supporting if the file was passed on the command line or detected as a supported config file.

Marking as resolved.